### DISCUSSION TITLE
LLaMa 3.2 vision - Clarification on User Input Masking During Training

### DISCUSSION THREAD

**Discussion by ronyadgar** [pytorch/torchtune#2878] [Q&A]
I'm noticing that during training, user input always appears to be masked. For context, you can refer to this section of the source code: https://github.com/pytorch/torchtune/blob/main/torchtune/data/_messages.py#L931C8-L931C53.

Could someone explain the reasoning behind this design? Specifically, I'm curious why the system doesn't support training on unmasked user input (where user input here refers strictly to text and not vision data). Any insights or further information on this design choice would be greatly appreciated.

**Comment by demirciberk** [comment]
Confirmed facts (from source and official docs):

1. Masking is controlled by mask_messages() in torchtune/data/_messages.py, based on a MaskingStrategy enum with three options: train_on_all, train_on_assistant (default), train_on_last.

2. The line you linked is inside the train_on_all branch:
python
elif masking_strategy == MaskingStrategy.TRAIN_ON_ALL:
    message.masked = message.role == "user" and message.contains_media

This means: even when you explicitly opt in to training on user text (masking_strategy="train_on_all", or the deprecated train_on_input=True), user messages that contain an image are still forced to masked=True. Plain text-only user messages are unmasked under train_on_all, as expected.

3. This is a deliberate design decision, not a bug. It's explicitly discussed in the maintainers' own RFC issue meta-pytorch/torchtune#2261 "Additional chat loss masking strategies", where a maintainer states:

   > "I think, we should be masking if the user input is multimodal even if train_on_all masking strategy is used?" — "Yes I agree with this."

   So the current behavior (any user message with contains_media=True is always masked, regardless of masking strategy) is the intended outcome of that discussion, implemented consistently across the built-in message transforms (InputOutputToMessages, ShareGPTToMessages, OpenAIToMessages, etc.).

4. Class docstring for InputOutputToMessages documents this explicitly:

masking_strategy (Optional[str]): ...
    - train_on_all: both user and assistant messages are unmasked
    - train_on_assistant: user messages are masked, only assistant messages are unmasked
    - train_on_last: only the last assistant message is unmasked

    Note: Multimodal user messages are always masked.

See message_transforms.rst and the Messages guide for the general masking model (Message.masked, tokenize_messages returning (tokens, mask)).

Hypothesis on the underlying rationale (not stated verbatim in torchtune docs, inferred from the architecture):

Llama 3.2 Vision fuses image embeddings into the text stream via cross-attention (VisionCrossAttentionMask), not by turning images into ordinary sequence tokens with a normal causal LM loss target. There's related discussion in meta-pytorch/torchtune#2493 about how the cross-attention mask sees the full text sequence (not causally restricted) at fusion time. Training a next-token loss directly over the region of the sequence where image content is interleaved is not a well-defined objective the same way plain text is (there's no "correct next token" prediction task for the placement of an image token, and the surrounding text was authored by the user, not generated by the model), so torchtune's authors chose to always exclude multimodal user turns from the loss to avoid an ill-posed/inconsistent training signal, independent of whatever masking strategy the user picks for plain text turns. If you need the exact motivating rationale beyond "we agreed it should be masked," that would require asking a maintainer directly or checking the PR history for the contains_media check (not linked in the current docs).

Practical takeaway: there is currently no supported config flag to unmask the text portion of a multimodal user turn while still masking the image content — the mask decision is per-Message, not per-token within a message's content list. If you want to train on user text in multimodal turns, you'd need to either (a) split the user turn so the image-only content stays masked and the accompanying text is issued as a separate unmasked message, or (b) subclass/override mask_messages/write a custom Transform (see Custom message transforms) that sets Message.masked per your own policy after contains_media is checked.
